### DISCUSSION TITLE
KC 18, ERR_TOO_MANY_REDIRECTS, error=expired_code

### DISCUSSION THREAD

**Discussion by AndreiMihaiRad** [keycloak/keycloak#12209] [Archive]
Hi, recently, I encounter a problem when running KC 18 on Kubernetes with Istio envoy proxy.
Currently, I have 2 replica counts, and I keep getting ERR_TOO_MANY_REDIRECTS just after I enter credentials on login screen.
On server side I see a lot of logs like following

` {"timestamp":"2022-05-26T15:05:10.344Z","sequence":8575,"loggerClassName":"org.jboss.logging.Logger","loggerName":"org.keycloak.events","level":"WARN","message":"type=LOGIN_ERROR, realmId=master, clientId=null, userId=null, ipAddress=10.240.245.5, error=expired_code, restart_after_timeout=true, authSessionParentId=6101cb6c-c9d8-448f-8bb3-e853cafd66cb, authSessionTabId=vRm1XzM0tCo","threadName":"executor-thread-1","threadId":86,"mdc":{},"ndc":"","hostName":"keycloak-feat-acc-3358-579955bf5b-6mzmg","processName":"QuarkusEntryPoint","processId":122}`

I tried to reduce replicas to just one instance, but the problem is still there.

If you have some suggestions for me I will really appreciate it.

**Comment by pedroigor** [comment]
Do you have the error response from the server when doing the code exchange?

**Comment by mpreu** [comment]
I have the same issue. After I run into the current documented problems to run KC behind a proxy at all (edge termination, --hostname-strict-https, etc.), running within an Istio Mesh only works for the landing page. Login is not possible with the same error as above.

The setup in my case is:

    Openshift Route (Edge Terminated) -> Istio Ingress Gateway Service -> Keycloak

Considering the general proxy related topics, this chain I am able to get working (including login, etc.):

    Openshift Route (Edge Terminated) -> Keycloak

I'll try to get some more detailed logs from the code exchange as requested if OP hasn't done it at that point.

**Reply by pedroigor** [reply]
So it works if you remove Istio from the chain?

The code expiration time is set using the `Client Login Timeout`. I'm wondering if you are using the default value (60s) or lower?

In any case, it does not make sense because it seems to work without Istio.

I'm asking for the server error response to see if we get more details about what is happening There are different places where `expired_code` can be returned not only for code expiration.

**Reply by ja1931** [reply]
@mpreu did you find the cause or resolution to this? I am having the same issue after running my Keycloak with Istio.

**Reply by ja1931** [reply]
Just in case this helps anyone else. It turned out the cache ports were being blocked by istio in my implementation. Set up authorization policy to allow it and my issues stopped.

**Comment by CameronNemo** [comment]
I am also seeing this error with Istio. I have not changed the login timeout from the default, and I am definitely logging in in less than a minute.

[example-org-json.txt](https://github.com/keycloak/keycloak/files/9017738/example-org-json.txt)

**Comment by pedroigor** [comment]
@CameronNemo Thanks. For some reason the `KC_RESTART` cookie is being set, indicating that the authentication session should be restarted. Sorry, but I'm not sure what might be causing this, although I see now where it is happening in the code.

I suspect about two things:

* For some reason the `tabId` is causing this because it is used to lookup the authentication session. But to properly check this we would need to enable debug or trace levels to see if we can spot something.
* I see `X-Firefox-Spdy` header in responses, and I think it indicates you are upgrading to http2. Can you check if whitout envoy the protocol is also http2?

Other than that, hard to say what might be happening and I'm trying to look at the differences between the two scenarios (with and without envoy).

**Comment by lanmark7** [comment]
I am trying to set up a basic Keycloak instance using Azure App Service connected to a keycloak instance.. after trying everything to get it to work I finally figured out that for some reason sessions needed to be sticky.. or in Azure App Service it is called "ARR Affinity" .. once  I turned this on everything started to work and the constant too many redirect issue is resolved.

I finally discovered that if I set the instance count to 1 everything works .. but by default app services on Azure have a instance count of two for obvious clustering reasons.

**Comment by pedroigor** [comment]
Hi All. Looks like this issue went a bit out of scope. If you think we have anything else to discuss, please re-open it.

**Comment by vijay-veeranki** [comment]
@AndreiMihaiRad  is the issue sorted now?

I get the same issue when i have STRICT istio and multiple keycloak replicas, If i just use one replica it is working fine though.

**Reply by AndreiMihaiRad** [reply]
yes, I did not encounter the problem any more. we can close issue.

**Comment by sreerajpaliyath** [comment]
Use Incognito window of Google Chrome browser or clear the cookies then run the keycloak login URL. it will work . Problem with the cookies.
